zfit/core/constraint.py

```diff
@@ -0,0 +1,191 @@
+#  Copyright (c) 2019 zfit
+
+import abc
+from collections import OrderedDict
+
+from typing import Dict, Union, Callable, Optional
+
+from .baseobject import BaseNumeric
+from .interfaces import ZfitParameter
+from ..util import ztyping
+from ..util.graph import get_dependents_auto
+from ..util.container import convert_to_container
+from ..util.exception import ShapeIncompatibleError, LogicalUndefinedOperationError
+from ..settings import ztypes
+from zfit import ztf
+import zfit
+
+import tensorflow as tf
+import numpy as np
+import tensorflow_probability as tfp
+
+tfd = tfp.distributions
+
+
+class BaseConstraint(BaseNumeric):
+
+    def __init__(self, params: Union[Dict[str, ZfitParameter]] = None,
+                 name: str = "BaseConstraint", dtype=ztypes.float,
+                 **kwargs):
+        """Base class for constraints.
+
+        Args:
+            dtype (DType): the dtype of the constraint
+            name (str): the name of the constraint
+            params (Dict(str, :py:class:`~zfit.Parameter`)): A dictionary with the internal name of the
+                parameter and the parameters itself the constrains depends on
+        """
+        super().__init__(name=name, dtype=dtype, params=params, **kwargs)
+
+    def value(self):
+        return self._value()
+
+    @abc.abstractmethod
+    def _value(self):
+        raise NotImplementedError
+
+    def _get_dependents(self) -> ztyping.DependentsType:
+        return self._extract_dependents(self.get_params())
+
+    def sample(self, n):
+        """Sample `n` points from the probability density function for the constrained parameters.
+
+        Args:
+            n (int, tf.Tensor): The number of samples to be generated.
+        Returns:
+            Dict(Parameter: n_samples)
+        """
+        sample = self._sample(n=n)
+        return {p: sample[:, i] for i, p in enumerate(self.get_params())}
+
+    @abc.abstractmethod
+    def _sample(self, n):
+        raise NotImplementedError
+
+
+class SimpleConstraint(BaseConstraint):
+
+    def __init__(self, func: Callable, params: Optional[ztyping.ParametersType] = None, sampler: Callable = None):
+        """Constraint from a (function returning a) Tensor.
+
+        The parameters are named "param_{i}" with i starting from 0 and corresponding to the index of params.
```

Comment:
Why the parameters need to be renamed?

Reply:
Sorry for the slow reaction!

Let's put it the other way: parameter.name as key is circular. Assuming that _whenever you have the name you also have the object -> no access with name strings_:
to get a parameter, you need the parameter itself then
```
param1 = constr.params[param1.name]
```
which is kinda useless. Furthermore, if you _don't_ know, you cannot access the parameter. But the param names should be the names of the _parametrization_, not of the actual value used. You get that by accessing the parametrization.

And `"param_{i}"` is just a generic name you can access easily programatically. But it's _actually_ not really needed or important.

Do you agree?

Reply:
Ok yes I agree :)